---
title: "Ouija: Interpretable marker-based single-cell pseudotime using Bayesian parametric models"
author: "Kieran R Campbell"
date: "July 2017"
output: BiocStyle::html_document
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
library(ggplot2)
theme_set(theme_bw())

set.seed(1L)

knitr::opts_chunk$set(echo = TRUE, cache = FALSE, 
                      message = FALSE, warning = FALSE,
                      fig.center = TRUE, fig.width = 6, fig.height = 4)
```

# Introduction

`Ouija` allows the incorporation of prior biological knowledge into single-cell trajectory learning in a principled, probabilstic manner. 

The input to Ouija should be a cell-by-gene expression matrix of non-negative values. We recommend using `log2(TPM + 1)` or `log2(RPKM + 1)` as this is what the mean-variance relationship in the model is designed for.

Under the hood, Ouija uses Bayesian non-linear factor analysis with priors on the factor loading matrix to specify gene behaviour. Inference is performed using the [Stan](http://mc-stan.org) probabilistic programming language.


# A basic example 

## Data retrieval

Here we can use some synthetic data bundled with the package. This contains a gene expression matrix `example_gex` comprising 11 genes and 400 cells, of which the first 9 are switch-like and the final 2 are transient

```{r load-synth-data}
devtools::load_all("~/oxford/bnlfa/oui/ouija")
data(example_gex)

example_gex[1:3, ]
```

## Response types

Using Ouija we can model genes as either exhibiting monotonic up or down regulation (known as switch-like behaviour), or transient behaviour where the gene briefly peaks. By default Ouija assumes all genes exhibit switch-like behaviour (and don't worry if you get it wrong - the noise model means incorrectly specifying a transient gene as switch-like has minimal effect).

In this example we can infer the behaviour type from the gene names:

```{r}
response_type <- sapply(strsplit(colnames(example_gex), "_"), `[`, 1)
```

## Fitting with Ouija

In order to fit the pseudotimes simply call `ouija` passing in the expected response types. Note that if no response types are provided then they are all assumed to be switch-like by default. The input to Ouija is either a cell-by-gene matrix of non-negative expression values, or an `ExpressionSet` that has similar values in `exprs(eset)`.

```{r ouija-fit, results = 'hide', message = FALSE, error = FALSE}
oui <- ouija(example_gex, response_type, iter = 1000)
```

```{r}
print(oui)
```


It's good practice to look at the trace and aurocorrelation of the (log)-likelihood to make sure the distribution has (roughly) converged. More advanced diagnostics may be accessed through the `rstan` package applied to `oui$fit`.

```{r plot-diagnostics}
plot_diagnostics(oui)
```

## Examining results

We can plot the gene expression against the MAP pseudotime too:

```{r plot-map-pseudotime}
plot_expression(oui)
```

```{r}
plot_switch_times(oui)
```


```{r}
plot_peak_times(oui)
```


## Determining regulation order of genes

```{r}
gene_regs <- gene_regulation(oui)
head(gene_regs)
```

```{r, fig.height = 10, fig.width = 6}
ggplot(gene_regs, aes(y = label, x = mean_difference, color = signif)) +
  geom_point() +
  geom_errorbarh(aes(xmin = lower_95, xmax = upper_95)) +
  xlab("Mean difference in regulation time") +
  ylab("Gene pair")
```



## Identifying metastable states

```{r}
plot_consistency(oui)
```

```{r}
cmo <- consistency_matrix(oui)
cell_classifications <- cluster_consistency(cmo)
```

```{r}
map_pst <- map_pseudotime(oui)
df_class <- data.frame(map_pst, cell_classifications)
ggplot(df_class, aes(x = map_pst, y = cell_classifications)) +
  geom_point() +
  xlab("MAP pseudotime") +
  ylab("Cell classification")
```

Alternatively

```{r}
ggplot(df_class, aes(x = map_pst, fill = factor(cell_classifications))) +
  geom_density() +
  theme(legend.position = 'top') +
  scale_fill_discrete(name = "Cell classification") +
  xlab("MAP pseudotime")
```


# Advanced usage

## Providing prior information on gene behaviour

## Manipulating the STAN fit

## Inference types

[Stan](http://mc-stan.org) now supports two types of inference: 

* Hamiltonian Monte Carlo (HMC) - full MCMC inference where gradient information of the log-posterior is used to "guide" the random walk through the parameter space
* Automatic Differentiation Variational Bayes (ADVI or simply VI) - approximate inference where the KL divergence to an approximate distribution is minimised

In general, HMC will provide more accurate inference with approximately correct posterior variance for all parameters. However, VB is orders of magnitude quicker than HMC and while it may underestimate posterior variance, anecdotally it seems just as good as HMC for discovering posterior pseudotimes. *If you are using Ouija on more than just a small panel of marker genes it is highly recommended to use Variational Bayes*.

These inference types may be invoked using the `inference_type` argument:

```{r inference-type, eval = FALSE}
oui_vb <- ouija(synth_gex, response_types,
                inference_type = "vb")
```



# Technical info

```{r sess-info}
sessionInfo()
```
